[clang-repl] Set up executor implicitly to account for init PTUs

weliveindetail opened the pull request:
Until now the IncrExecutor is created lazily on the first execution request. In order to process the PTUs that come from initialization, we have to do it upfront implicitly.

weliveindetail committed:
[clang-repl] Set up executor implicitly to account for init PTUs

weliveindetail committed:
[tmp] Add crash note

hahnjo commented on clang/test/Interpreter/inline-virtual.cpp:
```diff
@@ -14,7 +14,7 @@ struct A { int a; A(int a) : a(a) {} virtual ~A(); };
 
-// Create one instance with new and delete it.
+// Create one instance with new and delete it. We crash here now:
 A *a1 = new A(1);
```
Do we even have initial PTUs in the default case? Also the minimal reproducer shows a more general version where the `virtual` destructor is actually defined inline (c861d32d7c2791bdc058d9d9fbaecc1c2f07b8c7 addresses the case where it is out-of-line, which is special due to key `virtual` functions). So if that breaks entirely (which is critical for us), I'm personally not ok with just `XFAIL`ing it to land another change...

vgvassilev replied:
Yes, probably is something worth fixing now. I could not see the stack trace on osx. Can you paste it here?

weliveindetail replied:
> Do we even have initial PTUs in the default case?

Well, this test passes, if I comment out the for loop in the ctor that executes initial PTUs:
https://github.com/llvm/llvm-project/pull/84758/files#diff-b8484f1fc5b057f146ed5d9b6e2cd47c3f6f5ae879c7a0eee44f0a272581a88cR250-R254

> Also the minimal reproducer shows a more general version where the virtual destructor is actually defined inline (https://github.com/llvm/llvm-project/commit/c861d32d7c2791bdc058d9d9fbaecc1c2f07b8c7 addresses the case where it is out-of-line, which is special due to key virtual functions).

Oh that's a good note, I had not considered the difference yet and actually they have different backtraces. Eventually, they both reach the same VTablePtr code though.

> I could not see the stack trace on osx. Can you paste it here?

Here is a diff (inline left, out-of-line right):
<img width="2081" alt="Screenshot 2024-03-12 at 11 12 36" src="https://github.com/llvm/llvm-project/assets/7307454/41b64f92-83ea-4f82-b983-0b05e32ace42">

> So if that breaks entirely (which is critical for us), I'm personally not ok with just `XFAIL`ing it to land another change...

What breaks here is the parser and this patch doesn't even touch it. Not sure I am missing something, but it seems that it triggers a bug that always existed and just didn't show up so far.

weliveindetail replied:
Let me share some more observations.

We **execute** the initial PTU and not only parse it. It shouldn't make a difference for the parser right? Still, I wasn't sure, especially since we unusally **only parse** runtime PTUs. So I made some experiments:
```
    Init      Runtime      Dtor-def  |  New     Delete  Automatic
(1) Execute   Execute   Out-of-line  |  fails   -       fails
                             Inline  |  fails   -       once*

(2) Execute   Parse     Out-of-line  |  fails   -       fails
                             Inline  |  once*   fails   once

(3) Parse     Any       Out-of-line  |  works   works   works
                             Inline  |  works   works   works
```

Without `-O2` everything works.
Automatic = `A a1(1);`
`once` = Works exactly once (including shutdown)
`once*` = Same, but only if dtor has side-effects e.g. printf call

(1) is the behavior with this patch. (3) was the status-quo before this patch.

vgvassilev replied:
Thanks for the detailed analysis. @hahnjo can we “rent” some of your brain here to save us quite some debugging ;) ?

vgvassilev replied:
I remember we needed to do something to swap the tbaa analysis in codegen across ptus.

hahnjo replied:
I had a quick look here and got the same backtraces on Linux as @weliveindetail. From the debugger, it seems to fail because the DataLayout is not properly initialized? This could mean that executing the initial, empty PTU causes some damage in the internal data structures, but not sure...

vgvassilev replied:
Ah, does valgrind say anything?

vgvassilev replied:
Here is a report from valgrind and we seem to have a problem:

```
==858864== Warning: set address range perms: large range [0x108000, 0x11f2c000) (defined)
==858864== Invalid read of size 8
==858864==    at 0x11E99BE: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::length() const (basic_string.h:927)
==858864==    by 0x11F1E3B: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (basic_string.tcc:259)
==858864==    by 0x11EE135: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (basic_string.h:1387)
==858864==    by 0x11EB15C: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator=(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (basic_string.h:681)
==858864==    by 0x16F1819: llvm::DataLayout::operator=(llvm::DataLayout const&) (DataLayout.h:206)
==858864==    by 0x1A552E9: llvm::DataLayout::init(llvm::Module const*) (DataLayout.cpp:557)
==858864==    by 0x1A552B3: llvm::DataLayout::DataLayout(llvm::Module const*) (DataLayout.cpp:554)
==858864==    by 0x33797CA: clang::CodeGen::CodeGenTBAA::getVTablePtrAccessInfo(llvm::Type*) (CodeGenTBAA.cpp:274)
==858864==    by 0x3221BC3: clang::CodeGen::CodeGenModule::getTBAAVTablePtrAccessInfo(llvm::Type*) (CodeGenModule.cpp:1424)
==858864==    by 0x2DCA94E: clang::CodeGen::CodeGenFunction::InitializeVTablePointer(clang::CodeGen::CodeGenFunction::VPtr const&) (CGClass.cpp:2592)
==858864==    by 0x2DCAFC8: clang::CodeGen::CodeGenFunction::InitializeVTablePointers(clang::CXXRecordDecl const*) (CGClass.cpp:2676)
==858864==    by 0x2DC519C: clang::CodeGen::CodeGenFunction::EmitDestructorBody(clang::CodeGen::FunctionArgList&) (CGClass.cpp:1525)
==858864==  Address 0x13599270 is 512 bytes inside a block of size 800 free'd
==858864==    at 0x12D50B6F: operator delete(void*, unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==858864==    by 0x1E5A922: std::default_delete<llvm::Module>::operator()(llvm::Module*) const (unique_ptr.h:85)
==858864==    by 0x1E5A867: std::__uniq_ptr_impl<llvm::Module, std::default_delete<llvm::Module> >::reset(llvm::Module*) (unique_ptr.h:182)
==858864==    by 0x1E5A800: std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >::reset(llvm::Module*) (unique_ptr.h:456)
==858864==    by 0x1E4F8B8: std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >::operator=(decltype(nullptr)) (unique_ptr.h:397)
==858864==    by 0x1E4B5DC: llvm::orc::ThreadSafeModule::~ThreadSafeModule() (ThreadSafeModule.h:116)
==858864==    by 0x1F4DF4F: llvm::orc::IRMaterializationUnit::~IRMaterializationUnit() (Layer.h:31)
==858864==    by 0x1FCE1D5: llvm::orc::BasicIRLayerMaterializationUnit::~BasicIRLayerMaterializationUnit() (Layer.h:120)
==858864==    by 0x1FCE1F5: llvm::orc::BasicIRLayerMaterializationUnit::~BasicIRLayerMaterializationUnit() (Layer.h:120)
==858864==    by 0x1FCE235: std::default_delete<llvm::orc::BasicIRLayerMaterializationUnit>::operator()(llvm::orc::BasicIRLayerMaterializationUnit*) const (unique_ptr.h:85)
==858864==    by 0x1FCCFFD: std::unique_ptr<llvm::orc::BasicIRLayerMaterializationUnit, std::default_delete<llvm::orc::BasicIRLayerMaterializationUnit> >::~unique_ptr() (unique_ptr.h:361)
==858864==    by 0x1FCA870: llvm::orc::IRLayer::add(llvm::IntrusiveRefCntPtr<llvm::orc::ResourceTracker>, llvm::orc::ThreadSafeModule) (Layer.cpp:27)
==858864==  Block was alloc'd at
==858864==    at 0x12D4E013: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==858864==    by 0x2CC7B86: (anonymous namespace)::CodeGeneratorImpl::CodeGeneratorImpl(clang::DiagnosticsEngine&, llvm::StringRef, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, clang::HeaderSearchOptions const&, clang::PreprocessorOptions const&, clang::CodeGenOptions const&, llvm::LLVMContext&, clang::CoverageSourceInfo*) (ModuleBuilder.cpp:87)
==858864==    by 0x2CC90C0: clang::CreateLLVMCodeGen(clang::DiagnosticsEngine&, llvm::StringRef, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, clang::HeaderSearchOptions const&, clang::PreprocessorOptions const&, clang::CodeGenOptions const&, llvm::LLVMContext&, clang::CoverageSourceInfo*) (ModuleBuilder.cpp:372)
==858864==    by 0x2C8749E: clang::BackendConsumer::BackendConsumer(clang::BackendAction, clang::DiagnosticsEngine&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, clang::HeaderSearchOptions const&, clang::PreprocessorOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, clang::FileManager const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, llvm::SmallVector<clang::CodeGenAction::LinkModule, 4u>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >, llvm::LLVMContext&, clang::CoverageSourceInfo*) (CodeGenAction.cpp:127)
==858864==    by 0x2C8C2A0: clang::CodeGenAction::CreateASTConsumer(clang::CompilerInstance&, llvm::StringRef) (CodeGenAction.cpp:1051)
==858864==    by 0x22A1E64: clang::WrapperFrontendAction::CreateASTConsumer(clang::CompilerInstance&, llvm::StringRef) (FrontendAction.cpp:1207)
==858864==    by 0x229B7A5: clang::FrontendAction::CreateWrappedASTConsumer(clang::CompilerInstance&, llvm::StringRef) (FrontendAction.cpp:166)
==858864==    by 0x22A08F0: clang::FrontendAction::BeginSourceFile(clang::CompilerInstance&, clang::FrontendInputFile const&) (FrontendAction.cpp:955)
==858864==    by 0x2198EA6: clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (CompilerInstance.cpp:1061)
==858864==    by 0x2573ECA: clang::IncrementalParser::IncrementalParser(clang::Interpreter&, std::unique_ptr<clang::CompilerInstance, std::default_delete<clang::CompilerInstance> >, llvm::LLVMContext&, llvm::Error&) (IncrementalParser.cpp:211)
==858864==    by 0x2562095: std::_MakeUniq<clang::IncrementalParser>::__single_object std::make_unique<clang::IncrementalParser, clang::Interpreter&, std::unique_ptr<clang::CompilerInstance, std::default_delete<clang::CompilerInstance> >, llvm::LLVMContext&, llvm::Error&>(clang::Interpreter&, std::unique_ptr<clang::CompilerInstance, std::default_delete<clang::CompilerInstance> >&&, llvm::LLVMContext&, llvm::Error&) (unique_ptr.h:962)
==858864==    by 0x25599FD: clang::Interpreter::Interpreter(std::unique_ptr<clang::CompilerInstance, std::default_delete<clang::CompilerInstance> >, llvm::Error&) (Interpreter.cpp:236)
==858864== 
```

hahnjo replied:
That is with this patch or just current `main`?

vgvassilev replied:
Correct, that's also the case with releases/16.x...

vgvassilev replied:
https://github.com/llvm/llvm-project/pull/89031 should unblock this PR.

weliveindetail committed:
[wip] Fix unittests and highlight CreateJITBuilder() issue

The test for CreateJITBuilder() now fails with:
```
[ RUN      ] InterpreterExtensionsTest.CustomCrossJIT
/home/ez/Develop/llvm-project/clang/unittests/Interpreter/InterpreterExtensionsTest.cpp:254: Failure
Expected: (JIT) != (nullptr), actual: NULL vs (nullptr)
```

weliveindetail commented:
> https://github.com/llvm/llvm-project/pull/89031 should unblock this PR.

@vgvassilev It did. And it uncovered an issue that I didn't think about before: If we set up the executor in the constructor, then the interpreter instance isn't fully initialized yet when we create the LLJIT. Thus, we shouldn't call virtual functions on the interpreter yet (and if we do they will invoke the base-class methods and not the overidden ones). In particular, this breaks the `CreateJITBuilder()` mechanism: https://github.com/llvm/llvm-project/commit/0cf4788d9d0df60980cb48d28aafe7a86aa15a14

For me this patch doesn't have high prio, so I'd probably just drop it and keep everything as is for now. What's your take?

vgvassilev commented:
For me it is high priority because without it I cannot land the pch support in clang repl. I believe that’s also somewhat important for your use case?

weliveindetail commented:
Taking this patch would mean that deriving from Interpreter won't hold as the mechanism for extensions! At least for customizing the JITBuilder. I propose to pass the `LLJITBuilder` to the constructor then and to store it as a member for later `Create/ResetExecutor()` calls. Let me prototype that..

vgvassilev commented:
I am fine changing it in any way that creates an executor as part of the initialization phase when we have codegen. Maybe we can do it in the builder class and then pass it to the interpreter?

weliveindetail committed:
[clang][Interpreter] Take LLJITBuilder as an argument in the constructor

weliveindetail commented:
FYI: The run on the Windows bot failed even though tests worked as expected. In particular, all `ClangReplInterpreterTests` unittests passed as well as all `Interpreter` LIT tests. I could land this towards the end of the week.

vgvassilev commented:
@weliveindetail, I was wondering what's the fate of this patch?

weliveindetail commented:
@vgvassilev Thanks for the ping. I got derailed and missed the fix for the blocker!

weliveindetail commented:
Some bots report failures after this patch landed. I will push a fix soon.

weliveindetail commented:
@vgvassilev I pushed a quick fix for the tests. I think we should revisit/refactor the test story here though in the mid-term.

vgvassilev commented:
What do you mean?

vvereschaka commented:
Few bots are still failed because of the broken clang tests:
* https://lab.llvm.org/buildbot/#/builders/86/builds/80970
* https://lab.llvm.org/buildbot/#/builders/119/builds/18671
* https://lab.llvm.org/buildbot/#/builders/60/builds/17381

vgvassilev commented:
I am not sure how this patch changed these tests to start failing on Windows. Do you have any clue?

weliveindetail commented:
@vvereschaka Thanks for the note! should be fixed now.

> I am not sure how this patch changed these tests to start failing on Windows. Do you have any clue?

It's just cases I missed in the first quick-fix

> What do you mean?

This is adding a lot of boilerplate, because each and ever test now looks like this:
```
#ifdef CLANG_INTERPRETER_PLATFORM_CANNOT_CREATE_LLJIT
TEST(CodeCompletionTest, DISABLED_TemplateFunctions) {
#else
 TEST(CodeCompletionTest, TemplateFunctions) {
#endif
  if (!HostSupportsJit())
    GTEST_SKIP();
  <<< actual test >>>
}
```

@vgvassilev LIT tests check host JIT support during config and bulk skip all affected tests. (1) Maybe we can do sth like this in unittests as well? Or otherwise: (2) Add an option to the Interpreter to skip init PTUs and roll-back my quick fixes? What do you think?

vgvassilev commented:
Can we create some text fixture that will do that automatically for us and reduce this copy-paste?

weliveindetail commented:
Oh indeed, it seems we can decide to skip tests in `SetUp()`

```
class FooTest : public ::testing:Test {
protected:
    void SetUp()
    {
        GTEST_SKIP();
    }
};
```

jakeegan commented:
Hi, I see you fixed some tests, but there's still a couple failing on AIX:

>   Clang :: Interpreter/incremental-mode.cpp
>   Clang-Unit :: Interpreter/./ClangReplInterpreterTests/2/3
>   Clang-Unit :: Interpreter/ExceptionTests/./ClangReplInterpreterExceptionTests/0/1

https://lab.llvm.org/buildbot/#/builders/214/builds/12340/steps/6/logs/stdio

weliveindetail commented:
@jakeegan Thanks for reporting! Should be fixed now. Hope we caught them all 🤞

weliveindetail commented:
Here we go: https://github.com/llvm/llvm-project/pull/93816
